app/server.go

```diff
@@ -788,74 +788,100 @@ func (s *Server) DatabaseTypeAndMattermostVersion() (string, string) {
 	return *s.Config().SqlSettings.DriverName, mattermostVersion.Value
 }
 
-// initLogging initializes and configures the logger. This may be called more than once.
+// initLogging initializes and configures the logger(s). This may be called more than once.
 func (s *Server) initLogging() error {
+	// create the app logger if needed
 	if s.Log == nil {
-		s.Log = mlog.NewLogger(utils.MloggerConfigFromLoggerConfig(&s.Config().LogSettings, utils.GetLogFileLocation))
+		s.Log = mlog.NewLogger()
 	}
 
-	// Use this app logger as the global logger (eventually remove all instances of global logging).
-	// This is deferred because a copy is made of the logger and it must be fully configured before
-	// the copy is made.
-	defer mlog.InitGlobalLogger(s.Log)
-
-	// Redirect default Go logger to this logger.
-	defer mlog.RedirectStdLog(s.Log)
-
+	// create notification logger if needed
 	if s.NotificationsLog == nil {
-		notificationLogSettings := utils.GetLogSettingsFromNotificationsLogSettings(&s.Config().NotificationLogSettings)
-		s.NotificationsLog = mlog.NewLogger(utils.MloggerConfigFromLoggerConfig(notificationLogSettings, utils.GetNotificationsLogFileLocation)).
-			WithCallerSkip(1).With(mlog.String("logSource", "notifications"))
+		s.NotificationsLog = mlog.NewLogger().With(mlog.String("logSource", "notifications"))
 	}
 
+	// configure app and notification loggers
+	s.onLoggingConfigChange(nil, s.Config())
+
+	// Redirect default Go logger to app logger.
+	s.Log.RedirectStdLog(mlog.LvlStdLog)
+
+	// Use the app logger as the global logger (eventually remove all instances of global logging).
+	mlog.InitGlobalLogger(s.Log)
+
 	if s.logListenerId != "" {
 		s.RemoveConfigListener(s.logListenerId)
 	}
 	s.logListenerId = s.AddConfigListener(func(_, after *model.Config) {
-		s.Log.ChangeLevels(utils.MloggerConfigFromLoggerConfig(&after.LogSettings, utils.GetLogFileLocation))
-
-		notificationLogSettings := utils.GetLogSettingsFromNotificationsLogSettings(&after.NotificationLogSettings)
-		s.NotificationsLog.ChangeLevels(utils.MloggerConfigFromLoggerConfig(notificationLogSettings, utils.GetNotificationsLogFileLocation))
+		s.onLoggingConfigChange(nil, after)
 	})
+	return nil
+}
+
+// onLoggingConfigChange is called any time the configuration changes. All logging configuration is reloaded.
+func (s *Server) onLoggingConfigChange(_, after *model.Config) {
+	// configure main logger
+	if err := s.configureLogger("logging", s.Log, after.LogSettings, s.configStore, config.GetLogFileLocation); err != nil {
+		// revert to default logger if the config is invalid
+		mlog.InitGlobalLogger(nil)
+		mlog.Error("Error configuring logger", mlog.Err(err))
```

Comment:
Gentle reminder to check our logging [principles](https://developers.mattermost.com/contribute/server/style-guide/#log-levels) before merging this change.